src/v/cloud_roles/tests/schema_parsing_tests.cc

```diff
@@ -0,0 +1,139 @@
+}
+  )";
+
+    // schema is not a valid json (stray , at the end of the schema)
+    auto parse_schema_fail = parse(
+      schema_string, R"({"name": "John", "age":"42"})");
+    BOOST_CHECK(std::holds_alternative<cloud_roles::api_response_parse_error>(
+      parse_schema_fail));
+}
+
+auto to_str(cloud_roles::validate_and_parse_res const& val) {
+    return ss::visit(
+      val,
+      [](json::Document const&) -> ss::sstring { return "json::Document"; },
+      [](cloud_roles::malformed_api_response_error const& mal) {
+          return ssx::sformat(
+            "malformed_api_response_error: {}", mal.missing_fields);
+      },
+      [](cloud_roles::api_response_parse_error const& pe) {
+          return ssx::sformat("api_response_parse_error: {}", pe.reason);
+      });
+}
+
+BOOST_AUTO_TEST_CASE(schema_parsing_test) {
+    // rapidjson support for regex is just a subset of ecmascript, basically no
+    // character classes like \d or \w. when writing a pattern, fall back to a
+    // simple regex syntax
+    constexpr static auto schema_string = R"(
+{
+    "$schema": "http://json-schema.org/draft-04/schema#",
+    "type": "object",
+    "properties": {
+        "name": {
+            "type": "string"
+        },
+        "age": {
+            "type": "string",
+            "pattern": "^ *\\+?[0-9]{1,18} *$"
+        }
+    },
+    "required": ["name", "age"]
+}
+  )";
+
+    BOOST_TEST_CONTEXT("good data in") {
+        auto parse_ok = parse(
+          schema_string, R"({"name": "John", "age": "42"})");
+        BOOST_REQUIRE_MESSAGE(
+          std::holds_alternative<json::Document>(parse_ok), to_str(parse_ok));
+        auto& doc = std::get<json::Document>(parse_ok);
+        BOOST_CHECK(doc["name"].GetString() == std::string_view{"John"});
+        BOOST_CHECK(doc["age"].GetString() == std::string_view{"42"});
+    }
+
+    BOOST_TEST_CONTEXT("missing fields are reported") {
+        auto parse_missing = parse(schema_string, R"({"name": "John"})");
+        BOOST_CHECK_MESSAGE(
+          std::holds_alternative<cloud_roles::malformed_api_response_error>(
+            parse_missing),
+          to_str(parse_missing));
+        auto& doc = std::get<cloud_roles::malformed_api_response_error>(
+          parse_missing);
+        BOOST_CHECK(doc.missing_fields == std::vector<ss::sstring>{"age"});
+    }
+
+    BOOST_TEST_CONTEXT(
+      "missing fields takes precedence over malformed response (except for "
+      "rexexpr, 'feature' of rapidjson)") {
+        auto parse_missing_and_invalid = parse(
+          schema_string, R"({"age": "-42"})");
+        BOOST_CHECK_MESSAGE(
+          std::holds_alternative<cloud_roles::api_response_parse_error>(
+            parse_missing_and_invalid),
+          to_str(parse_missing_and_invalid));
+        // auto& doc = std::get<cloud_roles::malformed_api_response_error>(
+        //   parse_missing_and_invalid);
+        // BOOST_CHECK(doc.missing_fields == std::vector<ss::sstring>{"name"});
```

Comment:
why is this check commented out?

Reply:
old test i forgot to remove.

pattern:regex makes the rapidjson validation bail out before building the list of missing fields, so i had to change this section